src/tests/profiler/handles/handles.cs

```diff
@@ -0,0 +1,155 @@
+{
+    class TestClassForWeakHandle
+    {
+        public static int id;
+        public string Tag { get; set; }
+
+        public TestClassForWeakHandle(string tag)
+        {
+            Tag = $"{id}-{tag}";
+        }
+    }
+    class TestClassForStrongHandle
+    {
+        public static int id;
+        public string Tag = string.Empty;
+
+        public TestClassForStrongHandle(string tag)
+        {
+            Tag = $"{id}-{tag}";
+        }
+    }
+    class TestClassForPinnedHandle
+    {
+        public static int id;
+        public string Tag = string.Empty;
+
+        public TestClassForPinnedHandle(string tag)
+        {
+            Tag = $"{id}-{tag}";
+        }
+    }
+
+    class HandlesTests
+    {
+        class Objects
+        {
+            public TestClassForWeakHandle _weak;
+            public TestClassForStrongHandle _strong;
+            public TestClassForPinnedHandle _pinned;
+
+            // These objects will be allocated AFTER test objects.
+            // It is expected that after a compacting GC, they should be copied over
+            // previously allocated test objects that are no more referenced by handles.
+            public string _afterWeak;
+            public string _aferStrong;
+            public string _afterPinned;
+        }
+
+        static readonly Guid HandlesProfilerGuid = new Guid("A0F96622-522D-4654-AA56-BF421E79B210");
+
+
+        // The goal of this test is to validate the ICorProfilerInfo13 handle management methods:
+        //   CreateHandle (weak, strong, pinned)
+        //   DestroyHandle
+        //   GetObjectIDFromHandle
+        //
+        // SCENARIO:
+        //   1. Specific managed types instances are created but no reference are kept.
+        //   2. The corresponding native HandlesProfiler creates a handle for each.
+        //   3. A gen0 GC is triggered 
+        //   --> HandlesProfiler ensures:
+        //       - weak wrapped objects are no more alive
+        //       - strong and pinned wrapped objects are still alive
+        //   4. A gen1 is triggered.
+        //   5. HandlesProfiler destroys strong and pinned handles.
+        //   6. A gen2 is triggered.
+        //   7. HandlesProfiler ensures that no more instances are alive.
+        //
+        public static void DoWork(object parameter)
+        {
+            Objects objects = parameter as Objects;
+
+            AllocateInstances(objects);
+            Console.WriteLine($"weak = {objects._weak.Tag}");
+            objects._weak = null;
+            Console.WriteLine($"strong = {objects._strong.Tag}");
+            objects._strong = null;
+            Console.WriteLine($"weak = {objects._pinned.Tag}");
+            objects._pinned = null;
+
+            Console.WriteLine("Collection #0");
+            GC.Collect(0);
+
+            Console.WriteLine("Collection #1");
+            GC.Collect(1);
+
+            Console.WriteLine("Collection #2");
+            GC.Collect(2);
```

Comment:
It's not clear to me if you intend to only collect certain generations, the code makes it seem like you expect the argument to `GC.Collect` to be the count of the GCs so far. `GC.Collect(0)` will collect Gen 0, `GC.Collect(1)` will collect Gen 0 and Gen 1, etc.

Reply:
✅ This is definitively misleading  :^(
 I imagined that I would use the collected generation as an additional parameter but I did not used it. Updating the code to only collect gen0.